substrate/frame/treasury/src/tests.rs

```diff
@@ -217,16 +217,35 @@ impl Config for Test {
 	type BenchmarkHelper = ();
 }
 
-pub fn new_test_ext() -> sp_io::TestExternalities {
-	let mut t = frame_system::GenesisConfig::<Test>::default().build_storage().unwrap();
-	pallet_balances::GenesisConfig::<Test> {
-		// Total issuance will be 200 with treasury account initialized at ED.
-		balances: vec![(0, 100), (1, 98), (2, 1)],
+pub struct ExtBuilder {}
+
+impl Default for ExtBuilder {
+	fn default() -> Self {
+		Self {}
+	}
+}
+
+impl ExtBuilder {
+	pub fn build(self) -> sp_io::TestExternalities {
+		let mut t = frame_system::GenesisConfig::<Test>::default().build_storage().unwrap();
+		pallet_balances::GenesisConfig::<Test> {
+			// Total issuance will be 200 with treasury account initialized at ED.
+			balances: vec![(0, 100), (1, 98), (2, 1)],
+		}
+		.assimilate_storage(&mut t)
+		.unwrap();
+		crate::GenesisConfig::<Test>::default().assimilate_storage(&mut t).unwrap();
+		let mut ext = sp_io::TestExternalities::new(t);
+		ext.execute_with(|| System::set_block_number(1));
+		ext
+	}
+
+	pub fn build_ext_and_execute(self, test: impl FnOnce() -> ()) {
+		self.build().execute_with(|| {
+			test();
+			Treasury::do_try_state().expect("Storage invariants should hold.")
```

Comment:
can you also write individual tests to make sure `try-state` fails in each scenario it's supposed to?

Reply:
Are you suggesting to test the `try_state` implementation in the same way as the rest of the pallet? With the current implementation I don't see how that would work. As suggested here: https://github.com/paritytech/polkadot-sdk/issues/239 I'm running `try_state` after each regular test (as seen in the lines quoted by you). For example, if there was a test `fn try_state_invariant_1_works()`, which succeeds whenever invariant 1 is violated, then the following code would fail, because it leaves the storage in an invalid state:

```rust
try_state_invariant_1_works()  # <--- if this goes well...
Treasury::do_try_state().expect("Storage invariants should hold.") # <--- this will panick
```

A possible solution is to restore a valid state at the end of each test, but it seems to me that this  is a lot of overhead.

Reply:
in that case, I'd prefer removing the 'blindly checking' that try state invariants hold with more deliberate tests.

the blind tests *might* catch a bug but they probably won't and it doesn't really give us any assurances about our code. deliberate tests should make us feel much more comfortable that the code works as expected.

Reply:
roger that.